cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginInstantiator.java

```diff
@@ -218,14 +231,54 @@ private PluginProperties substituteMacroFields(Plugin plugin, MacroEvaluator mac
     for (Map.Entry<String, PluginPropertyField> pluginEntry : pluginPropertyFieldMap.entrySet()) {
       if (pluginEntry.getValue().isMacroSupported()) {
         String macroValue = plugin.getProperties().getProperties().get(pluginEntry.getKey());
-        properties.put(pluginEntry.getKey(), macroParser.parse(macroValue));
+        if (macroValue != null) {
+          macroValue = macroParser.parse(macroValue);
+        }
+        properties.put(pluginEntry.getKey(), macroValue);
+      } else {
+        properties.put(pluginEntry.getKey(), plugin.getProperties().getProperties().get(pluginEntry.getKey()));
+      }
+    }
+    return PluginProperties.builder().addAll(properties).build();
+  }
+
+  private PluginProperties substituteMacroDefaults(Plugin plugin) {
+    Map<String, String> properties = new HashMap<>();
+    Map<String, PluginPropertyField> pluginPropertyFieldMap = plugin.getPluginClass().getProperties();
+
+    TrackingMacroEvaluator trackingMacroEvaluator = new TrackingMacroEvaluator();
+    MacroParser macroParser = new MacroParser(trackingMacroEvaluator);
+
+    for (Map.Entry<String, PluginPropertyField> pluginEntry : pluginPropertyFieldMap.entrySet()) {
+      if (pluginEntry.getValue().isMacroSupported()) {
+        String macroValue = plugin.getProperties().getProperties().get(pluginEntry.getKey());
+        if (macroValue != null) {
+          macroParser.parse(macroValue);
+        }
+        if (trackingMacroEvaluator.hasMacro()) {
+          properties.put(pluginEntry.getKey(), getDefaultProperty(pluginEntry.getKey(),
+                                                                  pluginEntry.getValue().getType()));
+          trackingMacroEvaluator.reset();
+        } else {
+          properties.put(pluginEntry.getKey(), plugin.getProperties().getProperties().get(pluginEntry.getKey()));
+        }
       } else {
         properties.put(pluginEntry.getKey(), plugin.getProperties().getProperties().get(pluginEntry.getKey()));
       }
     }
     return PluginProperties.builder().addAll(properties).build();
   }
 
+  private String getDefaultProperty(String propertyName, String propertyType) {
+    Class propertyClass = propertyTypes.get(propertyType);
+    if (propertyClass == null) {
+      throw new IllegalArgumentException(String.format("Unable to get default value for property %s of type %s.",
+                                                 propertyName, propertyType));
+    }
+    return new InstantiatorFactory(false).get(TypeToken.of(Primitives.wrap(propertyClass))).create().toString();
```

Comment:
This is way to complicated to get the default value. Simply do:

```java
Object default = Defaults.defaultValue(propertyClass);
return default == null ? null : default.toString();
```

Since property type can be `String` and the default value of any object reference should be `null`, instead of empty string.